update authz docs for remote.ip (#8390)
istio/istio.io: content/en/docs/tasks/security/authorization/authz-ingress/index.md

An Istio authorization policy supports IP-based allow lists or deny lists as well as
the attribute-based allow lists or deny lists previously provided by Mixer policy.

> The Mixer policy is deprecated in *5 and not recommended for production use.

Reviewer: What is `*5`? Maybe:
```suggestion
Mixer policy has been deprecated since Istio 1.5 and is not recommended for production use.
```
Or is not even available in 1.8? If so, Maybe remove the sentence completely?

Author: That was a search/replace error.  I'll try to find out if Mixer is gone or not.

Author: mixer is gone, so i'll remove it.

    $ kubectl apply -f <(istioctl kube-inject -f @samples/httpbin/httpbin-gateway.yaml@) -n foo
    {{< /text >}}

* Turn on RBAC debugging in Envoy for the ingress gateway:

    {{< text bash >}}

>     $ kubectl get pods -n istio-system | grep ingress | awk '{print $1}' | while read -r pod; do istioctl proxy-config log "$pod" -n istio-system --level rbac:debug; done

Reviewer: This seems like a pretty crazy way to access the ingress gateway. Why not something like this?
```suggestion
    $ istioctl proxy-config log $(kubectl get pod -l istio=ingressgateway -n istio-system) -n istio-system --level rbac:debug
```

Author: You have to run istioctl proxy-config for each ingress gateway pod, so you can't escape doing a loop if there is more than one ingress gateway.  Also, you need to grab just the pod name output from get pods with cut or awk. I used awk.

Author: I made the other changes, and the remaining comments are related to this kind of loop.  Let me know if my answer is acceptable here, or if you still want me to rework it.

Reviewer: I see. The grep and awk still seem a little ugly. Would something like this work?
```
kubectl get pods -n istio-system -l istio=ingressgateway -o name | while ...
```

Author: That is close, but it prints the pods in this format `pod/istio-ingressgateway-99b56479c-46zt2` and istioctl doesn't accept the `pod/` part.  We could get rid of the grep that way, but the awk (or sed) will end up being a little more complex since a substitution will be necessary to remove `pod/`.  What do you think?

Reviewer: I don't know. Does this look any better?
```
kubectl get pods -n istio-system -l istio=ingressgateway -o name | sed 's|pod/||' | while ...
```
I'll leave it up to you.

Author: Sure, that seems reasonable.

    {{< text bash >}}
    $ kubectl patch svc istio-ingressgateway -n istio-system -p '{"spec":{"externalTrafficPolicy":"Local"}}'
    $ kubectl get pods -n istio-system | grep ingress | awk '{print $1}' | while read -r pod; do istioctl proxy-config log "$pod" -n istio-system --level rbac:debug; done
    {{< /text >}}

*  Follow the instructions in
    [Determining the ingress IP and ports](/docs/tasks/traffic-management/ingress/ingress-control/#determining-the-ingress-ip-and-ports)
    to define the `INGRESS_HOST` and `INGRESS_PORT` environment variables.

> 

Reviewer: You can't delete this, because you're using the env vars in the next step.

Author: Ah, I'll move that back up.

>     $ curl "$INGRESS_HOST":"$INGRESS_PORT"/headers -s -o /dev/null -w "%{http_code}\n"

Reviewer: A best practice in bash is to put quotes around the whole arg, not each variable:
```suggestion
    $ curl "$INGRESS_HOST:$INGRESS_PORT/headers" -s -o /dev/null -w "%{http_code}\n"
```

If you don’t see the expected output, retry after a few seconds.
Caching and propagation overhead can cause a delay.
{{< /warning >}}

## Getting traffic into Kubernetes and Istio

> * All methods of getting traffic into Kubernetes involve opening a port on all worker nodes.  The main features that accomplish this are the `NodePort` service and the `LoadBalancer` service.  Even the Kubernetes `Ingress` resource must be backed by an Ingress controller that will create either a `NodePort` or a `LoadBalancer` service.

Reviewer: This doesn't sound like it should be a bullet
```suggestion
All methods of getting traffic into Kubernetes involve opening a port on all worker nodes.  The main features that accomplish this are the `NodePort` service and the `LoadBalancer` service.  Even the Kubernetes `Ingress` resource must be backed by an Ingress controller that will create either a `NodePort` or a `LoadBalancer` service.
```

* A `NodePort` just opens up a port in the range 30000-32767 on each worker node and uses a label selector to identify which Pods to send the traffic to.  You have to manually create some kind of load balancer in front of your worker nodes or use Round-Robin DNS.

* A `LoadBalancer` is just like a `NodePort`, except it also creates an environment specific external load balancer to handle distributing traffic to the worker nodes.  For example, in AWS EKS, the `LoadBalancer` service will create a Classic ELB with your worker nodes as targets. If your Kubernetes environment does not have a `LoadBalancer` implementation, then it will just behave like a `NodePort`.  An Istio ingress gateway creates a `LoadBalancer` service.

> * What if the Pod that is handling traffic from the `NodePort` or `LoadBalancer` isn't running on the worker node that received the traffic?  Kubernetes has its own internal proxy called kube-proxy that receives the packets and forwards them to the correct node.

Reviewer: ```suggestion
What if the Pod that is handling traffic from the `NodePort` or `LoadBalancer` isn't running on the worker node that received the traffic?  Kubernetes has its own internal proxy called kube-proxy that receives the packets and forwards them to the correct node.
```

        numTrustedProxies: 1
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

For reference, here are the types of load balancers created by Istio with a `LoadBalancer` service on popular managed Kubernetes environments:

|Cloud Provider | Load Balancer Name            | Load Balancer Type
----------------|-------------------------------|-------------------
|AWS EKS        | Classic Elastic Load Balancer | TCP Proxy
|GCP GKE        | TCP/UDP Network Load Balancer | Network
|Azure AKS      | Azure Load Balancer           | Network
|DO DOKS        | Load Balancer                 | Network

{{< tip >}}
You can instruct AWS EKS to create a Network Load Balancer when you install Istio by using a `serviceAnnotation` like below:

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
  components:
    ingressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 5
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
{{< /text >}}

{{< /tip >}}

## IP-based allow list and deny list

> **`ipBlocks` vs `remoteIpBlocks`:** If you are using the X-Forwarded-For HTTP header or the Proxy Protocol to determine the original client IP address, then you should use `remoteIpBlocks` in your `AuthorizationPolicy`. If you are using `externalTrafficPolicy: Local`, then you should use `ipBlocks` in your `AuthorizationPolicy`.

Reviewer: In the `TCP/UDP Proxy Load Balancer` tab in the previous section, neither `X-Forwarded-For` nor `externalTrafficPolicy` is mentioned. Which one (ipBlocks or remoteIpBlocks) to use in that case?

This paragraph should also reference back to the previous section to make it clear that this decision is related to that information.

Author: I added a table here to make that more clear.

> 
> *  Follow the instructions in
>     [Determining the ingress IP and ports](/docs/tasks/traffic-management/ingress/ingress-control/#determining-the-ingress-ip-and-ports)
>     to define the `INGRESS_HOST` and `INGRESS_PORT` environment variables.

Reviewer: This needs to have been done already in "Before you begin".
```suggestion
```

kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: ALLOW
  rules:
  - from:
    - source:
        ipBlocks: ["1.2.3.4", "5.6.7.0/24"]
EOF
{{< /text >}}

{{< /tab >}}

{{< tab name="remoteIpBlocks" category-value="remoteIpBlocks" >}}

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: ALLOW
  rules:
  - from:
    - source:
        remoteIpBlocks: ["1.2.3.4", "5.6.7.0/24"]
EOF
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

* Verify that a request to the ingress gateway is denied:

    {{< text bash >}}
    $ curl "$INGRESS_HOST":"$INGRESS_PORT"/headers -s -o /dev/null -w "%{http_code}\n"
    403
    {{< /text >}}

* Update the `ingress-policy` to include your client IP address:

{{< tabset category-name="source" >}}

{{< tab name="ipBlocks" category-value="ipBlocks" >}}

Find your original client IP address if you don't know it and assign it to a variable:

{{< text bash >}}

> $ CLIENT_IP=$(kubectl get pods -n istio-system | grep ingress | awk '{print $1}' | while read -r pod; do kubectl logs "$pod" -n istio-system | grep remoteIP; done | head -1 | awk -F, '{print $3}' | awk -F: '{print $2}' | sed 's/ //') && echo "$CLIENT_IP"

Reviewer: Same question as https://github.com/istio/istio.io/pull/8390/files#r516295115

Author: Same reasoning here; we can't be sure which ingress gateway got the request in the case of multiple ingress gateway pods.  We need to loop through them all and print the logs to find the request.

192.168.10.15
{{< /text >}}

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: ALLOW
  rules:
  - from:
    - source:
        ipBlocks: ["1.2.3.4", "5.6.7.0/24", "$CLIENT_IP"]
EOF
{{< /text >}}

{{< /tab >}}

{{< tab name="remoteIpBlocks" category-value="remoteIpBlocks" >}}

Find your original client IP address if you don't know it and assign it to a variable:

{{< text bash >}}

> $ CLIENT_IP=$(kubectl get pods -n istio-system | grep ingress | awk '{print $1}' | while read -r pod; do kubectl logs "$pod" -n istio-system | grep remoteIP; done | head -1 | awk -F, '{print $4}' | awk -F: '{print $2}' | sed 's/ //') && echo "$CLIENT_IP"

Reviewer: https://github.com/istio/istio.io/pull/8390/files#r516295115

192.168.10.15
{{< /text >}}

Create the AuthorizationPolicy:

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: ALLOW
  rules:
  - from:
    - source:
        remoteIpBlocks: ["1.2.3.4", "5.6.7.0/24", "$CLIENT_IP"]
EOF
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

* Verify that a request to the ingress gateway is allowed:

    {{< text bash >}}

>     $ curl "$INGRESS_HOST":"$INGRESS_PORT"/headers -s -o /dev/null -w "%{http_code}\n"

Reviewer: ```suggestion
    $ curl "$INGRESS_HOST:$INGRESS_PORT/headers" -s -o /dev/null -w "%{http_code}\n"
```

    200
    {{< /text >}}

* Update the `ingress-policy` authorization policy to set
the `action` key to `DENY` so that the IP addresses specified in the `ipBlocks` are
not allowed to access the ingress gateway:

{{< tabset category-name="source" >}}

{{< tab name="ipBlocks" category-value="ipBlocks" >}}

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: DENY
  rules:
  - from:
    - source:
        ipBlocks: ["$CLIENT_IP"]
EOF
{{< /text >}}

{{< /tab >}}

{{< tab name="remoteIpBlocks" category-value="remoteIpBlocks" >}}

{{< text bash >}}
$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: ingress-policy
  namespace: istio-system
spec:
  selector:
    matchLabels:
      app: istio-ingressgateway
  action: DENY
  rules:
  - from:
    - source:
        remoteIpBlocks: ["$CLIENT_IP"]
EOF
{{< /text >}}

{{< /tab >}}

{{< /tabset >}}

* Verify that a request to the ingress gateway is denied:

    {{< text bash >}}

>     $ curl "$INGRESS_HOST":"$INGRESS_PORT"/headers -s -o /dev/null -w "%{http_code}\n"

Reviewer: ```suggestion
    $ curl "$INGRESS_HOST:$INGRESS_PORT/headers" -s -o /dev/null -w "%{http_code}\n"
```

    403
    {{< /text >}}

* You could use an online proxy service to access the ingress gateway using a
different client IP to verify the request is allowed.

* If you are not getting the responses you expect, view the ingress gateway logs which should show RBAC debugging information:

    {{< text bash >}}

>     $ kubectl get pods -n istio-system | grep ingress | awk '{print $1}' | while read -r pod; do kubectl logs "$pod" -n istio-system; done

Reviewer: https://github.com/istio/istio.io/pull/8390/files#r516295115

* A `NodePort` just opens up a port in the range 30000-32767 on each worker node and uses a label selector to identify which Pods to send the traffic to.  You have to manually create some kind of load balancer in front of your worker nodes or use Round-Robin DNS.

* A `LoadBalancer` is just like a `NodePort`, except it also creates an environment specific external load balancer to handle distributing traffic to the worker nodes.  For example, in AWS EKS, the `LoadBalancer` service will create a Classic ELB with your worker nodes as targets. If your Kubernetes environment does not have a `LoadBalancer` implementation, then it will just behave like a `NodePort`.  An Istio ingress gateway creates a `LoadBalancer` service.

* What if the Pod that is handling traffic from the `NodePort` or `LoadBalancer` isn't running on the worker node that received the traffic?  Kubernetes has its own internal proxy called kube-proxy that receives the packets and forwards them to the correct node.

## Source IP address of the original client

> If a packet goes through an external proxy load balancer and/or kube-proxy, then the original source IP address of the client is lost.  Below are some strategies for preserving the original client IP for logging or security purposes.  For reference, here are the types of load balancers created by Istio with a `LoadBalancer` service on popular managed Kubernetes environments:

Reviewer: I would move the `For reference, here ...` to after the tabset.
```
If a packet goes through an external proxy load balancer and/or kube-proxy, then the original source IP address of the client is lost. Here are some strategies for preserving the original client IP for logging or security purposes:

{{< tabset category-name="lb" >}}
...
{{< /tabset >}}

For reference, here are the types of load balancers created by Istio with a `LoadBalancer` service on popular managed Kubernetes environments:

|Cloud Provider | Load Balancer Name            | Load Balancer Type
----------------|-------------------------------|-------------------
|AWS EKS        | Classic Elastic Load Balancer | TCP Proxy
|GCP GKE        | TCP/UDP Network Load Balancer | Network
|Azure AKS      | Azure Load Balancer           | Network
|DO DOKS        | Load Balancer                 | Network

{{< tip >}}
You can instruct AWS EKS to create a Network Load Balancer when you install Istio by using a `serviceAnnotation` like below:

{{< text yaml >}}
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  meshConfig:
    accessLogEncoding: JSON
    accessLogFile: /dev/stdout
  components:
    ingressGateways:
    - enabled: true
      k8s:
        hpaSpec:
          maxReplicas: 10
          minReplicas: 5
        serviceAnnotations:
          service.beta.kubernetes.io/aws-load-balancer-type: "nlb"
{{< /text >}}

{{< /tip >}}
```

Reviewer: Yeah, tabset looks fine to me, as long as you move the "for reference" part after it.

* A `LoadBalancer` is just like a `NodePort`, except it also creates an environment specific external load balancer to handle distributing traffic to the worker nodes.  For example, in AWS EKS, the `LoadBalancer` service will create an ELB with your worker nodes as targets.  If your Kubernetes environment does not have a `LoadBalancer` implementation, then it will just behave like a `NodePort`.  An Istio ingress gateway creates a `LoadBalancer` service.

* What if the Pod that is handling traffic from the `NodePort` or `LoadBalancer` isn't running on the worker node that received the traffic?  Kubernetes has its own internal proxy called kube-proxy that receives the packets and forwards them to the correct node.

## Source IP address of the original client

* If a packet goes through an external load balancer and/or kube-proxy, then the original source IP address of the client is lost.  There are a few ways to handle this if you need the original client IP for logging or security purposes:

>       1. If you are using an HTTP/HTTPS external load balancer, it can put the original client IP address in the X-Forwarded-For header.  Istio can extract the client IP address from this header with some configuration.  See [Configuring Gateway Network Topology](/docs/ops/configuration/traffic-management/network-topologies/). Quick example if using a single load balancer in front of Kubernetes:

Reviewer 2: would it be better if we put the 3 options in a tab?

Reviewer: This one is better left as a numbered list, IMO.

Author: I made it a tab before I saw Frank's comment.  Let me know whether you like it this way or the other way.  I can change it back if need be.

## Source IP address of the original client

If a packet goes through an external proxy load balancer and/or kube-proxy, then the original source IP address of the client is lost.  Below are some strategies for preserving the original client IP for logging or security purposes.

{{< tabset category-name="lb" >}}

{{< tab name="TCP/UDP Proxy Load Balancer" category-value="proxy" >}}

If you are using a TCP/UDP Proxy external load balancer (AWS Classic ELB), it can use the [Proxy Protocol](https://www.haproxy.com/blog/haproxy/proxy-protocol/) to embed the original client IP address in the packet data.  Both the external load balancer and the Istio ingress gateway must support the proxy protocol for it to work.  In Istio, you can enable it with an `EnvoyFilter` like below:

{{< text yaml >}}
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: proxy-protocol
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
  - applyTo: LISTENER
    patch:
      operation: MERGE
      value:
        listener_filters:

>         - name: envoy.listener.proxy_protocol

Reviewer 3: ```
        - name: envoy.listener.tls_inspector
```

Is not needed?

Author: It's not needed just to enable the proxy protocol.  If you already have it in the config, the EnvoyFilter above won't remove it.  I'm not sure whether they are compatible or not though.

Reviewer 3: Without `envoy.listener.tls_inspector` `https://` does not work; maybe add it as note? Also its mentioned in here https://github.com/istio/istio.io/blob/master/content/en/docs/ops/configuration/traffic-management/network-topologies/index.md#proxy-protocol and in here https://github.com/istio/istio/pull/12748#issuecomment-640274792.

Author: I didn't notice they had also added Proxy protocol documentation to the Network Topologies doc, but you're right the two EnvoyFilters there and here should probably match.

Author: FYI, PR here: https://github.com/istio/istio.io/pull/8472

Reviewer 3: Perfect; thank you. Much appreciated.

      rules:
      - from:
        - source:
            ipBlocks: ["1.2.3.4", "5.6.7.0/24"]
    EOF
    {{< /text >}}

1. Here is the same policy, but using `remoteIpBlocks` instead of `ipBlocks`:

    {{< text bash >}}
    $ kubectl apply -f - <<EOF
    apiVersion: security.istio.io/v1beta1
    kind: AuthorizationPolicy
    metadata:
      name: ingress-policy
      namespace: istio-system
    spec:
      selector:
        matchLabels:
          app: istio-ingressgateway
      action: ALLOW
      rules:
      - from:
        - source:
            remoteIpBlocks: ["1.2.3.4", "5.6.7.0/24"]
    EOF

>     {{< /text >}}

Reviewer: Instead of making these remoteIpBlocks examples another step, you should put the two examples into a {{< tabset >}} in the previous step.
https://preliminary.istio.io/latest/about/contribute/shortcodes/#use-tabs

Author: Good point, I'll change it to a tab.

      rules:
      - from:
        - source:
            ipBlocks: ["1.2.3.4", "5.6.7.0/24", "$CLIENT_IP"]
    EOF
    {{< /text >}}

> 1. With `remoteIpBlocks`:

Reviewer 2: also probably use tab here, I think we want the user to choose one of {ipBlocks, remoteIpBlocks} but not both.

      rules:
      - from:
        - source:
            ipBlocks: ["$CLIENT_IP"]
    EOF
    {{< /text >}}

> 1. With `remoteIpBlocks`:

Reviewer 2: same here for tab.
